common-utils: Fix name resolution on IPv6-only hosts/networks

pjakma opened the pull request:
af_inet_client_get_remote_sockaddr is broken for IPv6 only
hosts/networks and hostnames.

It manually sets family to AF_INET6 if it detects the given hostname is 
actually an IPv6 address, but that leaves family set to AF_INET for 
other cases. Which means that for IPv6 host, the gf_resolve_ip6 call
 to getaddrinfo will fail - there will be no AF_INET addrs for a v6 only host.

Fix this by just getting rid of the attempt to set AF_INET6 for v6
addresses, and removing family specification.

Just use AF_UNSPEC in the getaddrinfo call to let getaddrinfo and
underlying address selections mechanisms do their thing.

Should work where the hostname is a v6 address too.

Fixes client mounting on v6-only.

pjakma committed:
common-utils: Fix name resolution on IPv6-only hosts/networks

af_inet_client_get_remote_sockaddr is broken for IPv6 only
hosts/networks and hostnames.

It manually sets family to AF_INET6 if it detects the hostname is an
IPv6 address, but that leaves family set to AF_INET for other cases.
Which means that for IPv6 host, the gf_resolve_ip6 call to getaddrinfo
will fail - there will be no AF_INET addrs for a v6 only host.

Fix this by just getting rid of the attempt to set AF_INET6 for v6
addresses, and removing family specification.

Just use AF_UNSPEC in the getaddrinfo call to let getaddrinfo and
underlying address selections mechanisms do their thing.

Should work where the hostname is a v6 address too.

Fixes client mounting on v6-only.

gluster-ant commented:
Can one of the admins verify this patch?

gluster-ant commented:
Can one of the admins verify this patch?

gluster-ant commented:
Can one of the admins verify this patch?

gluster-ant commented:
CLANG-FORMAT FAILURE:
Before merging the patch, this diff needs to be considered for passing clang-format

```diff --git a/libglusterfs/src/common-utils.c b/libglusterfs/src/common-utils.c
index f7b460248..a331d8c16 100644
--- a/libglusterfs/src/common-utils.c
+++ b/libglusterfs/src/common-utils.c
@@ -3325,7 +3325,7 @@ gf_process_reserved_ports(unsigned char *ports, uint32_t ceiling)
 out:
     GF_FREE(ports_info);
 
-#else /* FIXME: Non Linux Host */
+#else  /* FIXME: Non Linux Host */
     ret = 0;
 #endif /* GF_LINUX_HOST_OS */
 
@@ -5185,7 +5185,7 @@ close_fds_except_custom(int *fdv, size_t count, void *prm,
             closer(i, prm);
     }
     sys_closedir(d);
-#else /* !GF_LINUX_HOST_OS */
+#else  /* !GF_LINUX_HOST_OS */
     struct rlimit rl;
     int ret = -1;
 
```

pjakma committed:
nit: Remove unused variable

pjakma committed:
nit: fix clang-format issue from CI

Don't think I changed this, but the CI says I have to fix it.

pjakma commented:
Have addressed the clang-format nit (don't think I introduced that), removed the unused var, and looks like the CI passes. This is also running on my client - and is necessary to make it work there.

mohit84 commented:
/run regression

mohit84 commented:
> af_inet_client_get_remote_sockaddr is broken for IPv6 only
> hosts/networks and hostnames.
> 
> It manually sets family to AF_INET6 if it detects the given hostname is
> actually an IPv6 address, but that leaves family set to AF_INET for
> other cases. Which means that for IPv6 host, the gf_resolve_ip6 call
> to getaddrinfo will fail - there will be no AF_INET addrs for a v6 only host.
> 
> Fix this by just getting rid of the attempt to set AF_INET6 for v6
> addresses, and removing family specification.
> 
> Just use AF_UNSPEC in the getaddrinfo call to let getaddrinfo and
> underlying address selections mechanisms do their thing.
> 
> Should work where the hostname is a v6 address too.
> 
> Fixes client mounting on v6-only.

We do take family address decision based on transport.address-family value in volfile. If you change the address family to AF_UNSPEC, will getaddrinfo return always IPV6 address in case of ipv6 and IPV4 in case of IPV4.If the behavior of getaddrinfo is it may return IVP4 or IPV6 then we can;t go with this change. Can you please specify what is the exact issue you are facing?

pjakma commented:
The code I removed was taking the decision based on whether the string was an IPv6 address. Taking the decision based on an explicit configuration is another thing. But this code wasn't doing that. This code /required/ AF_INET for everything other than IPv6 addresses (in string form).

The problem I have is that Glusterfs appears not to work with IPv6-only hosts. 

1. Clients can not glusterfs mount IPv6-only servers. This patch should fix that.

2. Servers can not gluster peer probe IPv6-only new peers. I don't know if this patch fixes that - been waiting for stuff to compile over night to test that.

mohit84 commented:
> The code I removed was taking the decision based on whether the string was an IPv6 address. Taking the decision based on an explicit configuration is another thing. But this code wasn't doing that. This code /required/ AF_INET for everything other than IPv6 addresses (in string form).
> 
> The problem I have is that Glusterfs appears not to work with IPv6-only hosts.
> 
> 1. Clients can not glusterfs mount IPv6-only servers. This patch should fix that.
> 2. Servers can not gluster peer probe IPv6-only new peers. I don't know if this patch fixes that - been waiting for stuff to compile over night to test that.

By default IPV6 is disabled, i hope you have build the rpms after using the flag --with-ipv6-default  with configure. This check does not care about the transport-address-family value, it sets the protocol family to AF_INET if host name is ipv6. Can you share the error message you are getting while you are running peer probe.

pjakma commented:
IPv6 is enabled on Fedora. The error was that getaddrinfo was failing in 'mount -t glusterfs ...' on a client, trying to mount from servers with IPv6-only addresses for their hostname in DNS. The getaddrinfo call in gf_resolve_ip6 returns no records, and so af_inet_client_get_remote_sockaddr was printing the "DNS resolution failed on host" error.

[2021-07-27 09:51:48.179319 +0000] E [MSGID: 101075] [common-utils.c:519:gf_resolve_ip6] 0-resolver: error in getaddrinfo [{family=2}, {ret=Name or service not known}] 
[2021-07-27 09:51:48.179356 +0000] E [name.c:264:af_inet_client_get_remote_sockaddr] 0-glusterfs: DNS resolution failed on host ipv6-only-host.example.com

The reason why is that family was set to 2 - AF_INET. If you specify the v6 address it works. Looking more I guess it's cause client_fill_address_family has set family to 2 for some reason. Though I don't see how.

 I can't see anyway, with mount on a client, to specify transport on the commandline. I can't specify mount -t glusters -o transport.address-family=inet6 (the key client_fill_address_family looks for) anyway, that gives "Invalid option: transport.address-family".

Glusterfs is broken with IPv6-only-hostname servers, wrt mounting from clients at least.

I'm happy to get more pointers on where the right place is to go fix this, but the code as it stands appears to be broken, wrt "Just do the right thing" and IPv6.

pjakma commented:
My philosophy would be to keep things simple, and to let the underlying mechanisms already in the OS and the network, determine which addresses to use.

I.e., my tendency would be to simplify away a lot of the code in glusterfs around trying to work out which AF to use - certainly those based on heuristics around the form of the string - and just let the underlying platform (as administered by the sysadmins^WDevOps) take care of it.

mohit84 commented:
> IPv6 is enabled on Fedora. The error was that getaddrinfo was failing in 'mount -t glusterfs ...' on a client, trying to mount from servers with IPv6-only addresses for their hostname in DNS. The getaddrinfo call in gf_resolve_ip6 returns no records, and so af_inet_client_get_remote_sockaddr was printing the "DNS resolution failed on host" error.
> 
> [2021-07-27 09:51:48.179319 +0000] E [MSGID: 101075] [common-utils.c:519:gf_resolve_ip6] 0-resolver: error in getaddrinfo [{family=2}, {ret=Name or service not known}]
> [2021-07-27 09:51:48.179356 +0000] E [name.c:264:af_inet_client_get_remote_sockaddr] 0-glusterfs: DNS resolution failed on host ipv6-only-host.example.com
> 
> The reason why is that family was set to 2 - AF_INET. If you specify the v6 address it works. Looking more I guess it's cause client_fill_address_family has set family to 2 for some reason. Though I don't see how.
> 
> I can't see anyway, with mount on a client, to specify transport on the commandline. I can't specify mount -t glusters -o transport.address-family=inet6 (the key client_fill_address_family looks for) anyway, that gives "Invalid option: transport.address-family".
> 
> Glusterfs is broken with IPv6-only-hostname servers, wrt mounting from clients at least.
> 
> I'm happy to get more pointers on where the right place is to go fix this, but the code as it stands appears to be broken, wrt "Just do the right thing" and IPv6.

If you expect client should be able to mount the volume on IPV6 even without v6 address then you have to pass an argument to the client process (-otransport.address-family=inet6) at the time of mounting a volume. For specific to glusterd you can configure same in volfile.

mohit84 commented:
> My philosophy would be to keep things simple, and to let the underlying mechanisms already in the OS and the network, determine which addresses to use.
> 
> I.e., my tendency would be to simplify away a lot of the code in glusterfs around trying to work out which AF to use - certainly those based on heuristics around the form of the string - and just let the underlying platform (as administered by the sysadmins^WDevOps) take care of it.

If we do set AF_UNSPEC family my concern is are we sure 100% getaddrinfo will return correct address, if a user having both kind of environment then it may not work that's why gluster provides a configurable option to use any one (ipv4 or ipv6)

pjakma commented:
You can't set that on the mount options on a client though, AFAICT: 

# > /tmp/gl-debug.txt;  mount -o transport.address-family=inet6,log-file=/tmp/gl-debug.txt,log-level=DEBUG -t glusterfs v6host.example.com:/gv0 /mnt/Invalid option: transport.address-family

pjakma commented:
And the code should "Just Work" by default, with whatever addresses the admin has configured for their network. If a network has configured IPv6 addresses in hostnames, then that means they expect IPv6 to be used. It should not - by default - refuse to work with v6, and require further configuration to make it work.

Having options to /disable/ some AFs, for debug/whatever. Sure. But out of the box, it should just use whatever addresses the administrator(s) have configured for use on their network.

mohit84 commented:
> You can't set that on the mount options on a client though, AFAICT:
> 
> # > /tmp/gl-debug.txt; mount -o transport.address-family=inet6,log-file=/tmp/gl-debug.txt,log-level=DEBUG -t glusterfs v6host.example.com:/gv0 /mnt/Invalid option: transport.address-family

Kindly use like this -o xlator-option="transport.address-family=inet6"

pjakma commented:
That works, but it should not be necessary, should it?

mohit84 commented:
> That works, but it should not be necessary, should it?

As per current code it is necessary to retain the behavior of transport-address-family.

pjakma commented:
Right, but for client mounting there is no such option by default.

I think the client at least should just work. The servers comes with an obvious config file with the option commented out, making it clear.

pjakma commented:
There's also the question of whether it is necessary to disable IPv6 by default on the server.

mykaul commented:
@mohit84 - partially related - shouldn't we compile when we build RPMs with IPv6 enabled by default? (and if we do, add IPv6-specific tests!)

pjakma commented:
Oh, I'm perfectly willing to take guidance on the best way to fix this! The patch in this pull might be a hack - sure. ;)  I'd just like Gluster to be a bit more friendly to IPv6-only networking out of the box.
